dll/win32/shell32/folders/CFSFolder.cpp

```diff
@@ -244,6 +244,31 @@ getIconLocationForFolder(IShellFolder * psf, PCITEMID_CHILD pidl, UINT uFlags,
     return getDefaultIconLocation(szIconFile, cchMax, piIndex, uFlags);
 }
 
+// a callback for ExeFileHasAnyIcon
+static BOOL CALLBACK
+EnumResNameProc(
+    HMODULE hModule,
+    LPCWSTR lpszType,
+    LPWSTR lpszName,
+    LONG_PTR lParam)
+{
+    *(BOOL *)lParam = TRUE;
+    return FALSE;
+}
+
+static BOOL ExeFileHasAnyIcon(LPCWSTR pszFile)
```

Comment:
You can use `ExtractIconExW` for this, with a `-1` nIconIndex:
> Specifies the zero-based index of the first icon to extract. For example, if this value is zero, the function extracts the first icon in the specified file.
>
> **If this value is –1 and phiconLarge and phiconSmall are both NULL, the function returns the total number of icons in the specified file**. If the file is an executable file or DLL, the return value is the number of RT_GROUP_ICON resources. If the file is an .ico file, the return value is 1.

Reply:
I found our `ExtractIconExW` returns `1` even if exe has no icon. It seems we cannot use `ExtractIconExW` for this purpose.